Add daily_schedule integration

amitfin opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This is a new integration which should help using fixed automation rules, while controlling their behavior through this new entity type.
The "daily_schedule" entity has an on/off state based on the time periods provided as input by the user. It preserves the time periods during reboots.
A typical usage will be to create a static automation rule, which will behave according to changes in the time periods of the schedule, without the need to change the rule itself.
A demo clip can be found [here](https://www.youtube.com/watch?v=TluMAZpORwk).

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
daily_schedule:
  lights:
    name: Lights
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/15023

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-asc+-review%3Aapproved

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

amitfin committed:
Add "daily_schedule" integration

homeassistant commented:
Hi @amitfin, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#41385).

Once you do that we will be able to review and accept this pull request.

Thanks!

amitfin commented on homeassistant/components/input_timetable/translations/en.json:
```diff
@@ -0,0 +1,9 @@
+{
+    "state": {
+        "_": {
+            "off": "[%key:common::state::off%]",
```
This file was created by:
python3 -m script.translations develop
Based on the instructions at:
https://developers.home-assistant.io/docs/internationalization/core
However, I think that it's wrong, as the EN UI shows "[%key:common::state::off%]".
Should I fixed it manually by removing the "state" key altogether?

amitfin commented:
I've created a short video clip which demonstrates the end to end flow - https://www.youtube.com/watch?v=TluMAZpORwk
It will be great if someone can help with the code review (and later on the merge). Thanks!!

dgomes commented on CODEOWNERS:
```diff
@@ -221,6 +221,7 @@ homeassistant/components/input_datetime/* @home-assistant/core
 homeassistant/components/input_number/* @home-assistant/core
 homeassistant/components/input_select/* @home-assistant/core
 homeassistant/components/input_text/* @home-assistant/core
+homeassistant/components/input_timetable/* @home-assistant/core
```
You should add @amitfin not home-assistant/core

amitfin replied:
Done

dgomes commented on homeassistant/components/input_timetable/__init__.py:
```diff
@@ -0,0 +1,370 @@
+            vol.Any(
+                {
+                    vol.Optional(CONF_NAME): cv.string,
+                    vol.Optional(CONF_ICON): cv.icon,
```
We shouldn't make icon configurable, that is up to customisation or Lovelace

amitfin replied:
Done

dgomes reviewed (commented):
What does this accomplish, that you can't accomplish with an automation, or that it is much harder to accomplish with an automation ?

amitfin commented:
> What does this accomplish, that you can't accomplish with an automation, or that it is much harder to accomplish with an automation ?

@dgomes, this is an input element for automation rules which simplify the automation configuration for non-tech people (e.g. my 12 years old kid). It has a similar purpose like the other input elements (input_datetime, input_number, input_boolean, etc').

There are 2 main advantages for this element, compared to writing the data directly in the automation rule:

1) it's possible to control the automation rule behavior with the simplified (dedicated) UI for this element. See also this short video clip: https://www.youtube.com/watch?v=TluMAZpORwk

2) It's possible to create a single and simpler rule for on and off, by using the state of the element. Here is an example:
input_timetable:
  lights:
automation:
  - alias: Lights
    trigger:
      platform: state
      entity_id: input_timetable.lights
    action:
      - service: "switch.{{ 'turn_on' if is_state('input_timetable.lights', 'on') else 'turn_off' }}"
        data:
          entity_id: all
Without this element, the rule would have been much more complicated, with a list of times to turn the lights on, and another rule to turn them off.

We've been using this element for our household and it made the configuration accessible to the rest of the family. The config file (as an example) which we use can be found here: https://github.com/amitfin/Home-Assistant-Config/blob/master/automations/timer-automations.yaml As can be seen in this example, the rule itself is quite complicated. However, with the separation of the input_timetable, it's possible for the rest of the family members to control the actual time by just editing the input_timetable (in its dedicated UI).

The documentation of this new element (which also tried to explain the usage and motivation) can be found here: https://github.com/home-assistant/home-assistant.io/pull/15023/files?short_path=bdc0d59#diff-bdc0d599c3dc1ebb8643b16e4041196fce6ad653c9d7af41c5a840d5363dc3e0

Thoughts?

amitfin commented:
@dgomes, were there any offline discussions around this integration, or is it stale?
Do you see a fundamental difference between this automation rules' input element and the rest of the input elements (e.g. input_datetime, input_number, input_boolean, etc')?

dgomes commented:
Hi!

I'm not aware of any offline discussions. Your video is quite good at putting forward the use case for this integration, but still it takes more contributors to bring in their approval. 

Since the backlog is very large, I ask for your patience

sim-san commented:
Maybe you can use some logic or frontend code from this custom component...
https://community.home-assistant.io/t/scheduler-card-custom-component/217458

amitfin commented:
@sim-san, I really hope that at some point this integration will be reviewed and merged. I'll be happy to discuss further development of the frontend code, if it happens.

giggo commented:
> Maybe you can use some logic or frontend code from this custom component...
> https://community.home-assistant.io/t/scheduler-card-custom-component/217458

Yes. This one is phantastic. I don't get it why HA does not have any week/scheduler component by now.... This one should become the default.

Aohzan commented:
It should be great if you can choose the day(s) of the week associated with the time.

Also, a simpler automation example would be just like:
`  - alias: Clean
    trigger:
      platform: state
      entity_id: input_timetable.vacuum
    action:
      - service: vacuum.start
        data:
          entity_id: vacuum.my_vacuum`

amitfin commented:
> It should be great if you can choose the day(s) of the week associated with the time.

This is definitely supported. The new input_timetable is only one of the blocking elements of an automation rule. You can easily combine it with other elements, like the day(s) of the week. 
Here is the additional condition which will run the rule only on a certain day (there is nice UI to do it as well):
```
condition: time
weekday:
  - sun
```
![image](https://user-images.githubusercontent.com/19599059/114623963-a02ab100-9cb8-11eb-9af9-01b6685ef87e.png)

Aohzan commented:
> > It should be great if you can choose the day(s) of the week associated with the time.
> 
> This is definitely supported. The new input_timetable is only one of the building blocks of automation rules. You can easily combine it with other elements, like the day(s) of the week.
> Here is the additional condition which will run the rule only on a certain day:
> 
> ```
> condition: time
> weekday:
>   - sun
> ```
> 
> And here is the UI for that:
> ![image](https://user-images.githubusercontent.com/19599059/114623963-a02ab100-9cb8-11eb-9af9-01b6685ef87e.png)

Yes I know, but the condition apply to all triggers of time table, if I want to trigger an automation for every monday at 9, tuesday and wednesday at 12, I have to do two automations

amitfin commented:
@Aohzan, you are correct that different timetables for different days in the week requires multiple automation rules. On the other hand, adding days of week (or days of month, etc') will make the user experience of the common use case more complicated. Therefore, the decision was to keep this building block only about time of the day, and to reuse other building blocks for expressing days conditions (when needed).

Aohzan commented:
@amitfin ok I understand totally, thanks :+1:

MarkusHornbach commented:
@amitfin I'm really looking forward to the integration of your input_timetable, because this is exactly what I need for my climate scheduling, very nice work! 
One question: is it possible to use it as a custom component before official integration?

amitfin commented:
@MarkusHornbach, thanks! 
Yes, you can use it as a custom component. All you need to do is to copy the files in the PR under homeassistant/components/input_timetable on to HA at /config/custom_components/input_timetable (and restart HA).
In addition, you need to use a custom card, which I created as an interim solution, found at https://github.com/amitfin/Home-Assistant-Config/blob/master/www/timetable-card.js . It should be placed in the same path (/config/www/timetable-card.js). Follow the instruction at https://developers.home-assistant.io/docs/frontend/custom-ui/registering-resources to register it.
A sample card will look like this:
```
title: Timers
toggle: input_boolean.timers
entities:
  - entity: input_timetable.timer1
    name: light1
  - entity: input_timetable.timer2
    name: light2
```
I've been using this interim solution for months (while waiting for someone to review this component, please?)

amitfin commented:
@MarkusHornbach, please note that there was a recent breaking change, and there is a need to add a version when using custom integration (but version is not used for built-in integrations). See also [this](https://github.com/home-assistant/core/pull/49726).
So, basically, the following line need to be added manually to homeassistant/components/input_timetable/manifest.json:
"version": "1.0.0"
According to the announcement, this must be done before the next release (2021.6), which is supposed to come out in the coming Wed (Jun/2nd).

I know that it's becoming less and less convenient, but I don't know of anything that can help with the review process, except of being patient. It's sitting in the queue for 10 months, so hopefully we are getting there.

MarkusHornbach commented:
@amitfin Thanks for the hint to add the version!
So review and later official integration can only be done by the HA admins? 
What if you first go the way through HACS component? Or can your integration already be found there?

amitfin commented:
@MarkusHornbach, correct, only HA's collaborators have write access (merge the PR).

Regarding HACS: 
IMHO, eventually this component should be part of HA. The main reason is the user-experience. For example, it's not possible to add a helper element to the UI without changing the frontend codebase (e.g. https://github.com/home-assistant/frontend/pull/7687).
The reason for not using HACS as an interim solution is because I don't know if there is an easy path to migrate HACS component to a built-in component. It's possible to give them different names, but it means a manual migration.

I'm keeping my optimism that it will be reviewed soon...

clayauld commented:
> 
> 
> Thanks @clayauld! I've been using it for a year now, and it made our configuration and ease of use much better. My kids are controlling the timers, instead of me doing automation rules' modifications. I think that the code (and testing) is more or less ready, but I was unable getting help from the core team to review and eventually merge it. It's sitting in the queue for a while now. I'm not sure if there is a way to get some help here. Will be happy to hear an advice from anyone with more experience on how to make it happen.

One possibility in the interim is to create a HACS custom integration. I have created a frontend HACS card and have imported custom components outside of that for my use as well.

This would allow you to share the working code, allowing others to use it and provide feedback/bugfixes/improvements, and maintain it yourself while waiting to see if it gets merged into the mainline code.

If you're interested, let me know. I'd be happy to help you put that together and get it added to the HACS repositories for easy installation.

amitfin commented:
@clayauld, I'm using the greatest and latest version (currently 2021.10.2). Things are working smoothly on my system.

Here are the relevant files:

find config/custom_components/input_timetable
config/custom_components/input_timetable/__init__.py
config/custom_components/input_timetable/translations
config/custom_components/input_timetable/translations/en.json
config/custom_components/input_timetable/services.yaml
config/custom_components/input_timetable/manifest.json
config/custom_components/input_timetable/strings.json

(Note: you should add "version": "1.0.0" to "manifest.json")

find config/www/timetable-card.js            
config/www/timetable-card.js

(I'm pinging you over email, to help in troubleshooting.)

chrismanivong reviewed (approved):
Reviewed, checked out, integrated, working!

nivnoach commented on homeassistant/components/input_timetable/__init__.py:
```diff
@@ -0,0 +1,350 @@
+        self._config = config
+        self._update_state()
+
+    def _schedule_update(self) -> None:
```
this seems like very specific logic, can maybe add a bit of doc?

amitfin replied:
Done

nivnoach commented on tests/components/input_timetable/test_init.py:
```diff
@@ -0,0 +1,641 @@
+"""The tests for the input timetable component."""
```
neat coverage!

nivnoach reviewed (approved):
Looks like this component is in really good shape, and looks very useful! I wonder why not merge it and have in GA, considering the positive feedback?

danielbrunt57 commented:
@amitfin: Is this on you to correct??
```
Run . venv/bin/activate
  . venv/bin/activate
  python --version
  pylint homeassistant/components/{default_config,input_timetable}
  shell: bash --noprofile --norc -e -o pipefail {0}
  env:
    CACHE_VERSION: 5
    PIP_CACHE_VERSION: 1
    DEFAULT_PYTHON: 3.9
    PRE_COMMIT_CACHE: ~/.cache/pre-commit
    PIP_CACHE: /tmp/pip-cache
    SQLALCHEMY_WARN_20: 1
    PYTHONASYNCIODEBUG: 1
Python 3.9.0
************* Module homeassistant.components.input_timetable
Warning: homeassistant/components/input_timetable/__init__.py:108:22: W0020: Argument 1 should be of type HomeAssistant (hass-argument-type)
Error: Process completed with exit code 4.

danielbrunt57 commented:
@amitfin - Is this on you to correct??
```
Run . venv/bin/activate
  . venv/bin/activate
  python --version
  pylint homeassistant/components/{default_config,input_timetable}
  shell: bash --noprofile --norc -e -o pipefail {0}
  env:
    CACHE_VERSION: 5
    PIP_CACHE_VERSION: 1
    DEFAULT_PYTHON: 3.9
    PRE_COMMIT_CACHE: ~/.cache/pre-commit
    PIP_CACHE: /tmp/pip-cache
    SQLALCHEMY_WARN_20: 1
    PYTHONASYNCIODEBUG: 1
Python 3.9.0
************* Module homeassistant.components.input_timetable
Warning: homeassistant/components/input_timetable/__init__.py:108:22: W0020: Argument 1 should be of type HomeAssistant (hass-argument-type)
Error: Process completed with exit code 4.
```

amitfin commented:
@danielbrunt57, thanks! The lint error was fixed.
(The change from HomeAssistantType to HomeAssistant was introduced by https://github.com/home-assistant/core/pull/48400 on Mar 27, 2021, while this PR was created 5 months earlier on Oct 7, 2020. I've missed this change, which is now also checked by lint.)

danielbrunt57 commented:
@dgomes 
Hi Dave,

Just curious if we will see this approved/merged anytime soon...

Thanks,
Daniel

dgomes commented:
> @dgomes Hi Dave,
> 
> Just curious if we will see this approved/merged anytime soon...
> 
> Thanks, Daniel

Sorry, but I don't any estimate on that...

moskovskiy82 commented:
What is the difference? You can just put  an input date time and a switch to the frontend? Looks mor like a dedicated lovelace card

amitfin commented:
This is the plural version of input_datetime. It has a list of on/off time events. It allows creating a static automation rule which is controlled by this new helper element (via modifications to the time events list).

moskovskiy82 commented:
> This is the plural version of input_datetime. It has a list of on/off time events. It allows creating a static automation rule which is controlled by this new helper element (via modifications to the time events list).

It doesn't seem "plural" as it is just a switch with a time and date. You can easily create an automation with a switch in the condition part of the rule. What is the difference?

danielbrunt57 commented:
> > This is the plural version of input_datetime. It has a list of on/off time events. It allows creating a static automation rule which is controlled by this new helper element (via modifications to the time events list).
> 
> It doesn't seem "plural" as it is just a switch with a time and date. You can easily create an automation with a switch in the condition part of the rule. What is the difference?

Please show me an automation with a condition based on a switch which emulate this integration and would permit user manipulation of a variable list of times & dates via the frontend UI.

moskovskiy82 commented:
> > > This is the plural version of input_datetime. It has a list of on/off time events. It allows creating a static automation rule which is controlled by this new helper element (via modifications to the time events list).
> > 
> > 
> > It doesn't seem "plural" as it is just a switch with a time and date. You can easily create an automation with a switch in the condition part of the rule. What is the difference?
> 
> Please show me an automation with a condition based on a switch which emulate this integration and would permit user manipulation of a variable list of times & dates via the frontend UI.

This integration just checks constantly if it's in the period? Like let's say a heater which should be on from 7 to 21 and then off. In HA now you just set the rules which switch the heater on at 7 and off at 21. Unless you reboot your HA at 7 and it doesn't switch on at this time?

danielbrunt57 commented:
> > > > This is the plural version of input_datetime. It has a list of on/off time events. It allows creating a static automation rule which is controlled by this new helper element (via modifications to the time events list).
> > > 
> > > 
> > > It doesn't seem "plural" as it is just a switch with a time and date. You can easily create an automation with a switch in the condition part of the rule. What is the difference?
> > 
> > 
> > Please show me an automation with a condition based on a switch which emulate this integration and would permit user manipulation of a variable list of times & dates via the frontend UI.
> 
> This integration just checks constantly if it's in the period? Like let's say a heater which should be on from 7 to 21 and then off. In HA now you just set the rules which switch the heater on at 7 and off at 21. Unless you reboot your HA at 7 and it doesn't switch on at this time?

This integration provides far more versatility and ease of use for a non-HA administrative user...

![image](https://user-images.githubusercontent.com/49846893/163236622-f68395cc-16f5-4f7c-b81a-c1166293c443.png)

![image](https://user-images.githubusercontent.com/49846893/163236732-585df781-eca9-4913-a025-56e1869ae2a0.png)

![image](https://user-images.githubusercontent.com/49846893/163237005-eaca418e-804f-4239-84c6-aaca2e0af5b8.png)

danielbrunt57 commented:
@amitfin  - what are the conflicts??

amitfin commented:
> @amitfin - what are the conflicts??

Fixed. 
The format of CODEOWNERS, which is a metadata of the codebase, was slightly modified. No impact or change to the code of the component itself.

amitfin commented:
@frenck, please let me know if I can help. Otherwise, I won't do any further changes without an explicit ask. Thanks!

amitfin commented:
Thank you very much Franck! It's clear that you spent a significant amount of time on this topic, and thank you for the detailed summary!

Let me try contributing to the core team discussion:
1) (I’m probably stating the obvious, but anyway) This component is not about functionality. HA already supports the same functionality by editing the automation rules directly. This is true for this component as well as calendar and all of the “input_” components (and others). This is not a statement about the importance of this component (and the others) but about its purpose. The goal is: **better user experience and ease of use**.
2) There are daily routines, weekly routines, and ad-hoc routines (e.g., future events). It’s possible to create a component that can support all, but it will lose its purpose of providing a good user experience. For example, using a weekly schedule for daily routines is cumbersome as every day is configured separately (unless we make the day-of-week optional.) On a side note, it will be great to have a local calendar, as there are use cases which are best fit to be solved with a calendar. 
3) The schema of the configuration and the UI should have the same capabilities, or they can get out of sync. For example, if the configuration can specify arbitrary time, but the UI has 30 minutes granularity, it will be impossible to open the UI correctly if the configuration has a non 30 minutes entry.
4) Specifying a time range for “on” periods has advantages over a list of events. However, it brings an annoying corner case of a time range which is across days (e.g., starts at 10pm and ends at 2am). In addition, IMHO, time range overlapping has the potential to make things complicated (including the user experience). Anyway, I don’t have a strong opinion whether the advantages are greater than the disadvantages.
5) Just for context about the original motivation: this component is mimicking the old timers we used to have before the IoT period (and many people still do) like [this](https://www.amazon.com/dp/B07GBK39LG). 

More tactical points:
1) Naming: how about “daily_schedule” (or “weekly_schedule”)?
2) Agree about removing the "advanced" services. Is it ok to keep one – “reconfig”?

Thanks again! 
Amit.

frenck commented:
> There are daily routines, weekly routines, and ad-hoc routines (e.g., future events). It’s possible to create a component that can support all, but it will lose its purpose of providing a good user experience

We have no interest in that currently. We think a weekly schedule covers most cases for now and is a good start. Daily or ad-hoc schedules we like to keep out of scope. Even if you wanted daily schedules, defining a weekly one to achieve that is possible and not over complicated.

> The schema of the configuration and the UI should have the same capabilities, or they can get out of sync.

Yes, but since the StorageCollection is used, that shouldn't be an issue.

> For example, if the configuration can specify arbitrary time, but the UI has 30 minutes granularity, it will be impossible to open the UI correctly if the configuration has a non 30 minutes entry.

That is a UI discussion. We do not have 30 minute granularity. And nothing is impossible.

> Specifying a time range for “on” periods has advantages over a list of events. However, it brings an annoying corner case of a time range which is across days (e.g., starts at 10pm and ends at 2am). In addition, IMHO, time range overlapping has the potential to make things complicated (including the user experience).

I don't see the issue? That can be validated and protected against. As the ranges are defined per day, there won't be ranges across days. Ranges overlapping is, within each day, fairly easy to validate 🤷 

> Just for context about the original motivation: this component is mimicking the old timers we used to have before the IoT period (and many people still do) like [this](https://www.amazon.com/dp/B07GBK39LG).

Yes, and that is not what we are looking for. We are more looking for what is shown in the Node-RED screenshot above.

> Naming: how about “daily_schedule” (or “weekly_schedule”)?

I would just do `schedule`; it might get extended in the future. Calling it `weekly_schedule` prevents that.

> Agree about removing the "advanced" services. Is it ok to keep one – “reconfig”?

That should not be needed. Please remove them all first, keeping `reload` is fine though (as that is part of the storage collections.

amitfin commented:
This component was implemented for simplifying the control of our daily routines. Controlling a daily routine with a weekly schedule means editing the time periods 7 times (once per day), which makes it irrelevant as it doesn't simply the control, but make it harder, error prone, and tedious.
I understand that daily routines are not considered for now. It will be interesting to validate the assumption that weekly routines are more common than daily routines (it's definitely not the case for us.)

frenck commented:
> I understand that daily routines are not considered for now. It will be interesting to validate the assumption that weekly routines are more common than daily routines (it's definitely not the case for us.)

That is easy to validate, as you can make daily routines with a weekly schedule. So this will help both people looking for daily and weekly.

Yeah, agreed, daily users will have to repeat it. Something we can explore in the future (e.g., by supporting daily schemas, or maybe a UI feature to sync days up, or even a workday/weekend schedule thing). As for the YAML configuration, this could already be simplified with anchors.

amitfin commented:
> you can make daily routines with a weekly schedule

The purpose of this helper component is to make the configuration of daily routines more convenient. It doesn't enable a new use case, as everything is already possible by editing the automation rules themselves. Using a weekly schedule for daily routines doesn't make the configuration more convenient, so it's missing the purpose.

Don't get me wrong, I understand that core team doesn't think that daily routines are important enough for a dedicated helper component. I just don't know how to change this component in a way that will also address the use case that it was built for (and is being used for over the last 2 years) - making the configuration of daily routines more convenient.

Will you be open for 2 helper components: "weekly_schedule" and "daily_schedule"? It might sound redundant at first. But since the purpose here is making things convenient, having a dedicated component and simplified UI for each use case can make sense. If this is a possibility, I can commit to build both components. (I think that we agree on how to build these two.)

frenck commented:
> The purpose of this helper component is to make the configuration of daily routines more convenient.

In that case, we are heading to an endpoint. We have no interest in adding a daily schedule at this point in time.

> Don't get me wrong, I understand that core team doesn't think that daily routines are important enough for a dedicated helper component.

I (or we) didn't say that. We just want to propose starting with weekly and we can think/look/decide into other schedules in the future.

> Will you be open for 2 helper components

We don't need 2 helpers for that, as already stated above as well.

../Frenck

amitfin commented:
Ok. I understand.
I'll work on the improvements, based on the great feedback. Thanks!
When time allows, I'll try to work on the weekly version to contribute to this amazing community (regardless if it will be useful for us or not.)

Just as a closing thought: the main difference here is in the high-level approach. The proposal is a more powerful element to cover different use cases at the cost of the user experience (at least for the daily scenarios). I think that the platform (automation rule) itself already supports any type of scheduling, so the purpose of the "schedule" elements is purely around user experience. Therefore, having dedicated experiences for the different use cases is addressing better the need and goal. It's not a trivial decision though, as I understand the need to keep HA in a reasonable size, for obvious reasons. It's also against our instinct as engineers to try and generalize the solutions.

frenck commented:
? I totally surprised you adjusted this to a daily schedule? I think I've been pretty clear on that we won't be accepting that?

danielbrunt57 commented:
@frenck I am not opposed to a weekly version but how should we / @amitfin  proceed? I like this daily integration but I think your weekly integration concept will work too,  I'm just an end user and don't not want to see this die. Happy to help however I can....

amitfin commented:
The changes that were made are based on the feedback:
1) Removed the redundant advanced services.
2) Changed the configuration to be time periods (start and end) instead of event based (time and state).

I understand that we are still at a point that this component doesn't have the blessing of the core team. A weekly schedule will be accepted, and a daily schedule won't (according to the recent discussions). The name change to "daily_schedule" didn't change the logic or the purpose of this component. It was done to better describe it. 
It also allows me building a "weekly_schedule" (or just "schedule" if this is the name that will be chosen) as a contribution to the community.

I'm not trying to argue the core decision. This component was built for a purpose and is being used quite successfully. I can't change it to a weekly schedule and still use it for the same use case it was built for. If there was such an option, I would be (way) more than happy to take this path.

frenck commented:
> It also allows me building a "weekly_schedule" (or just "schedule" if this is the name that will be chosen) as a contribution to the community.

Please read again all my comments above, as you missed quite some parts of it apparently.
We will not be accepting a daily schedule, nor do we want to different integrations for weekly and daily.

amitfin commented:
Let me list a few use cases that this component is being used for:
1) Controlling the garden's watering system.
2) Controlling the attic's air ventilation device.
3) Controlling the swimming pool pump.
In all these cases we make modifications from time to time (when it becomes hotter, when the water in the swimming pool needs additional filtering, etc').

@frenck, can you please help me to understand how the suggested approach can address these use cases with a good user experience? I'm probably missing something here.

Regardless, I'm planning to fork this PR and adjust it to a weekly schedule, so eventually we will converge. (This PR will be the one that will be modified to weekly.)

frenck commented:
> Let me list a few use cases that this component is being used for:

We have considered a lot of use cases, and as I have motivated and explained above a couple of times now, this can be done using a weekly schedule as well.

> @frenck, can you please help me to understand how the suggested approach can address these use cases with a good user experience? I'm probably missing something here.

See all of the above. This has been discussed and responded to a couple of times now, I feel like we are running in circles 😟 

> Regardless, I'm planning to fork this PR and adjust it to a weekly schedule, so eventually we will converge. (This PR will be the one that will be modified to weekly.)

Please modify this PR and not create a second integration (as been indicated and told in previous comments as well).

danielbrunt57 commented:
@amitfin So, you cannot morph this from daily to weekly to meet @frenck's requirements and not break current users? If I am to understand all of the above, you need to maintain this current daily_schedule repository (which will not be merged with HA) to keep existing users running and create a new forked repository that will become the new weekly "schedule" version which you will then submit a new PR for,  thus this PR needs to end.  Am I correct?

amitfin commented:
A few changes were made based on the feedback and additional research, mainly looking at the built-in [Times of the Day](https://www.home-assistant.io/integrations/tod/) integration:
1) The integration creates a binary_sensor instead of an entity in its own domain.
2) The setup is through the UI instead of the yaml configuration file.
3) Configuring the schedule (time periods) is done through the UI. There is no need for an additional UI component (although a dedicated UI can be nicer.)

Here is a short [video clip](https://youtu.be/GvN6g0azPfk) demoing the new implementation.

Note: if you are copying the files manually to `custom_components`:
1) Don’t forget to add `"version": "1.0.0"` to `manifest.json`. 
2) This version is not backward compatible, so the configuration needs to be set from scratch.

frenck commented:
Closing this PR, as it is being replaced by #76566.

As mentioned above, we are currently not looking at adding a daily schedule. Unfortunately, this PR didn't adjust to that.

Nevertheless, thanks for being willing to contribute 👍 

../Frenck
